clinica/utils/filemanip.py

```diff
@@ -1,209 +1,476 @@
+
+    Raises
+    ------
+    FileNotFoundError
+        If the provided file does not exist.
+    """
+    from clinica.utils.filemanip import _zip_unzip_nii  # noqa
+
+    return _zip_unzip_nii(in_file, unzip_nii, same_dir)
 
-    if (in_file is None) or isinstance(in_file, _Undefined):
-        return None
 
-    if not isinstance(in_file, str):  # type(in_file) is list:
-        return [unzip_nii(f) for f in in_file]
+def save_participants_sessions(
+    participant_ids: List[str],
+    session_ids: List[str],
+    out_folder: str,
+    out_file: Optional[str] = "participants.tsv",
+) -> None:
+    """Save the participants and sessions to TSV.
 
-    _, base, ext = split_filename(in_file)
+    Parameters
+    ----------
+    participant_ids: List[str]
+        List of participant IDs.
 
-    # Not compressed
-    if ext[-3:].lower() != ".gz":
-        return in_file
-    # Compressed
-    gunzip = Gunzip(in_file=in_file)
-    gunzip.run()
-    return gunzip.aggregate_outputs().out_file
+    session_ids: List[str]
+        List of session IDs.
 
+    out_folder: str
+        Folder where the resulting TSV file should be written.
 
-def save_participants_sessions(participant_ids, session_ids, out_folder, out_file=None):
-    """Save <participant_ids> <session_ids> in <out_folder>/<out_file> TSV file."""
+    out_file: str, optional
+        Name of the TSV file. Default="participants.tsv".
+
+    Raises
+    ------
+    ValueError
+        If provided `participant_ids` and `session_ids` do not have
+        the same length.
+    """
     import os
+    from pathlib import Path
 
-    import pandas
+    import pandas as pd
 
     from clinica.utils.stream import cprint
 
-    assert len(participant_ids) == len(session_ids)
+    if len(participant_ids) != len(session_ids):
+        raise ValueError(
+            "The number of participant IDs is not equal to the number of session IDs."
+        )
 
+    out_folder = Path(out_folder)
     os.makedirs(out_folder, exist_ok=True)
-
-    if out_file:
-        tsv_file = os.path.join(out_folder, out_file)
-    else:
-        tsv_file = os.path.join(out_folder, "participants.tsv")
-
+    tsv_file = out_folder / out_file
+
+    data = pd.DataFrame(
+        {
+            "participant_id": participant_ids,
+            "session_id": session_ids,
+        }
+    )
     try:
-        data = pandas.DataFrame(
-            {
-                "participant_id": participant_ids,
-                "session_id": session_ids,
-            }
-        )
         data.to_csv(tsv_file, sep="\t", index=False, encoding="utf-8")
     except Exception as e:
         cprint(msg=f"Impossible to save {out_file} with pandas", lvl="error")
         raise e
 
 
+def _raise_non_bids_or_caps_compliant_filename(filename: str) -> None:
+    raise ValueError(
+        f"Input filename {filename} is not in a BIDS or CAPS compliant format."
+        " It does not contain the subject and session information."
+    )
```

Comment:
I looked at usages of this function and found it used twice, in L232 and L304. In both, the pattern is the same, i.e. a regex match on the path raising an exception if nothing found. I believe it's the whole thing that should be factored out, not just the raising an exception part.

Reply:
Actually, the pattern is slightly different, but I agree it is still worth a factorization at the cost of a small additional argument to this function. I also changed the name since it needs to return the match to the caller and not only raise.
Implemented in [7cb8c5a](https://github.com/aramis-lab/clinica/pull/781/commits/7cb8c5a4a578a0ba7ea0376176e8c151a7730198)